gazelle/gazelle.go

```diff
@@ -0,0 +1,258 @@
+// the appropriate language-specific equivalent) for each import according to
+// language-specific rules and heuristics.
+func (*bzlLibraryLang) Resolve(c *config.Config, ix *resolve.RuleIndex, rc *repo.RemoteCache, r *rule.Rule, importsRaw interface{}, from label.Label) {
+	imports := importsRaw.([]string)
+
+	r.DelAttr("deps")
+
+	if len(imports) == 0 {
+		return
+	}
+
+	deps := make([]string, 0, len(imports))
+	for _, imp := range imports {
+		if strings.HasPrefix(imp, "@") {
+			// This is a dependency that is external to the current repo.
+			deps = append(deps, strings.TrimSuffix(imp, fileType))
+		} else {
+			res := resolve.ImportSpec{
+				Lang: languageName,
+				Imp:  imp,
+			}
+			matches := ix.FindRulesByImport(res, languageName)
+
+			if len(matches) == 0 {
+				log.Printf("%s: %q was not found in dependency index. Skipping. This may result in an incomplete deps section and require manual BUILD file intervention.\n", from.String(), imp)
+			}
+
+			for _, m := range matches {
+				deps = append(deps, m.Label.String())
+			}
+		}
+	}
+
+	sort.Strings(deps)
+	if len(deps) > 0 {
+		r.SetAttr("deps", deps)
+	}
+}
+
+var kinds = map[string]rule.KindInfo{
+	"bzl_library": {
+		NonEmptyAttrs:  map[string]bool{"srcs": true, "deps": true},
+		MergeableAttrs: map[string]bool{"srcs": true},
+	},
+}
+
+// GenerateRules extracts build metadata from source files in a directory.
+// GenerateRules is called in each directory where an update is requested
+// in depth-first post-order.
+//
+// args contains the arguments for GenerateRules. This is passed as a
+// struct to avoid breaking implementations in the future when new
+// fields are added.
+//
+// A GenerateResult struct is returned. Optional fields may be added to this
+// type in the future.
+//
+// Any non-fatal errors this function encounters should be logged using
+// log.Print.
+func (*bzlLibraryLang) GenerateRules(args language.GenerateArgs) language.GenerateResult {
+	var rules []*rule.Rule
+	var imports []interface{}
+	for _, f := range append(args.RegularFiles) {
+		if ignoreSuffix.Matches(f) {
+			continue
+		}
+		if strings.HasSuffix(f, fileType) {
+			name := strings.TrimSuffix(f, fileType)
+			r := rule.NewRule("bzl_library", name)
+
+			r.SetAttr("srcs", []string{f})
+
+			if args.File == nil || !args.File.HasDefaultVisibility() {
+				var inPrivateDir bool
+				for _, d := range strings.Split(args.Dir, string(filepath.Separator)) {
+					if d == "private" {
+						inPrivateDir = true
+					}
+				}
+				if !inPrivateDir {
+					r.SetAttr("visibility", []string{"//visibility:public"})
+				}
+			}
+
+			fullPath := filepath.Join(args.Dir, f)
+			loads, err := getBzlFileLoads(fullPath)
+			if err != nil {
+				log.Printf("%s: contains symtax errors: %v", fullPath, err)
+				// Don't `continue` since it is reasonable to create a target even
+				// without deps.
+			}
+
+			rules = append(rules, r)
+			imports = append(imports, loads)
+		}
+	}
+	return language.GenerateResult{
```

Comment:
You may also want to populate `Empty`. For each `bzl_library` rule in `args.File` that only has `srcs` that aren't in `args.RegularFiles` or `args.GenFiles`, add a `bzl_library` with no `srcs` or `deps`. That will let Gazelle delete `bzl_library` rules after the corresponding `.bzl` files are deleted.